File: Leetcode-75/206-Reverse-Linked-List/reverse-linked-list.ts
// Problem 206 - REVERSE A LINKED LIST

// Given the head of a singly linked list, reverse the list, and return the reversed list.

// ---------------------------------------------------------------------------------------------------------------------------

import {
    ListNode,
    linkedListToList,
    listToLinkedList,
} from "../../Utils/TypeScript/LinkedList";

function reverseList(head: ListNode | null): ListNode | null {
    if (!head) return null; // Base case --> If not given a valid head, return null

    let prevNode: ListNode | null = null; // variable for the previous node --> Will become the tail
    let currNode: ListNode | null = head; // variable for the current node --> Tracks position within list

    while (currNode) {
        let nextNode = currNode.next; // variable for the next node in the list --> In the original order
        currNode.next = prevNode; // Change the next pointer to point to the previous node --> The REVERSAL is here
        prevNode = currNode; // Update variable to current node for next iteration
        currNode = nextNode; // Update variable to the next node for next iteration
    }

    return prevNode; // Return the previous node --> After the last iteration, the previous node represents the final current node AKA the new HEAD
}

//---------------------------------------------------------------------------------------------------------------------------

const linkedList1 = listToLinkedList([5, 4, 3, 2, 1]);
const linkedList2 = listToLinkedList([56, 34, 32, 4424, 544, 20, 4]);

const reversedList1 = reverseList(linkedList1);
const reversedList2 = reverseList(linkedList2);

console.log(linkedListToList(reversedList1));
console.log(linkedListToList(reversedList2));

File: Leetcode-75/206-Reverse-Linked-List/reverse-linked-list.py
# Problem 206 - REVERSE A LINKED LIST

# Given the head of a singly linked list, reverse the list, and return the reversed list.

# ---------------------------------------------------------------------------------------------------------------------------

from Utils.Python.LinkedList import ListNode, linkedListToList, listToLinkedList


def reverseList(head: ListNode) -> ListNode:
    # Variable for the previous node --> Will become the tail
    prev = None

    # Variable for our current position in our list --> Begins at the head
    currNode = head

    # While loop for moving through our nodes of our list
    while currNode:
        # Variable for the next node in the list originally
        nextNode = currNode.next

        # Make the pointer to the next node point to the previous node
        currNode.next = prev

        # Make the previous node pointer point to the current node --> Updated for next loop iteration
        prev = currNode

        # Updated the current node pointer to the next node --> For the next iteration of the loop
        currNode = nextNode

    return prev  # Return the previous node --> At the end of the loop, this becomes the new head of the linked list (updates to currNode AKA the new HEAD)


# ---------------------------------------------------------------------------------------------------------------------------


# SLIGHT SYNTAX CHANGE --> SAME SOLUTION (ONE LESS VAR STORED)
def reverseListSmall(head: ListNode) -> ListNode:
    if not head or not head.next:
        return head

    prev = None

    while head.next:
        next = head.next
        head.next = prev
        prev = head
        head = next

    head.next = prev

    return head


# ---------------------------------------------------------------------------------------------------------------------------

linkedList1 = listToLinkedList([1, 2, 3, 4, 5])
linkedList2 = listToLinkedList([56, 34, 32, 4424, 544, 20, 4])


reversedLinkedList1 = reverseList(linkedList1)  # Expect [5, 4, 3, 2, 1]
reversedLinkedList2 = reverseList(linkedList2)  # Expect [4, 20, 544, 4424, 32, 34, 56]


print(linkedListToList(reversedLinkedList1))
print(linkedListToList(reversedLinkedList2))
